# 2.7 异常处理

## Exception 与 Error

在学习 **异常处理** 之前，我们需要先分清，什么是 **异常**，什么是 **错误**

### 异常 Exception

异常通常指程序运行时出现的问题，而且是可以预测到并且可以修复的问题，比如文件读取时可能就会产生 `IOException`，或者数组越界产生的 `ArrayIndexOutOfBoundsException`

### 错误 Error

错误通常指程序运行中出现的严重错误，比如运行程序的时候，**JVM** 的内存溢出，就会产生 `OutOfMemoryError`

## try...catch...(finally...)

`try...catch...` 是 **Java** 中常用于捕获异常的结构，`try {}` 下用于执行可能会出现异常的代码，比如数据库操作，`catch()` 中填写异常类 `Exception` 和异常变量 `e`，当然，两者都可以变动，最后 `finally {}` 下填写的是不论异常是否被捕获，都会被执行的代码，比如，数据库操作中，不论数据是否写入成功，都需要关闭数据库连接，把操作写入 `finally {}` 的方法是首选

```Java
try {
    // 可能会出现异常的代码
} catch(Exception e) {
    // 异常被捕获时的操作
    e.printStackTrace(); // 输出异常
} finally {
    // 默认执行完毕 try...catch... 后最终执行的代码
}
```

## 抛出 throws / throw

你可以联系上面 `try...catch...` 的代码来理解 `throws`，可以看见，`throws` 可以省略掉 `try...catch...` 结构，可以直接写可能会出现异常的代码

```Java
public void throwException() throws IOException {
    // 可能会出现异常的代码
    ...
    saveConfig(); // 通常文件存储操作需要被 try...catch... 环绕，但我们已经抛出了这个异常，就无需使用 try...catch...
}
```

`throw` 相当于自己创建一个异常，当程序执行到这里时，抛出某种错误

```Java
public void throwException() throws Exception {
    throw new Exception("this is a exception");
}
```